torch/jit/__init__.py

```diff
@@ -1129,6 +1129,16 @@ def whichmodule(obj):
             pass
     return '__main__'
 
+def _recursive_compile_class(obj, loc):
+    _qual_name = _qualified_name(obj)
+    # We're starting a new compilation, so update the error call stack in
+    # case it fails
+    error_stack = torch._C.CallStack(_qual_name)
```

Comment:
I see only 1 call from python, so what is the rationale for moving the error stack logic into Python here? It exposes more API surface via duplication, and isn't handling exceptions correctly (that del won't run if compile_and_register_class fails, which is common).

Reply:
`error_stack` will still get deleted even in an exception, because it is locally scoped. i added the `del` to force python garbage collection handling but it's probably unnecessary, will delete.

I don't really like going from try_ann_to_type -> C++ -> python, it makes it harder to debug. I would prefer it was in python along with a lot of the other recursive compilation code. There's one call from python and one call from C++.

Reply:
The behavior with the local stack is an implementation detail of cpython, its not really robust to rely on it. If someone refactors this code and puts a loop or if statement in, it would be easy to get the stack messed up.

Reply:
If you want to avoid ping-ponging between python and C++,  then really `try_ann_to_type` should be moved to C++instead. It is already being called from C++ directly in a few places, increasing the ping-ponging. It requires transporting SourceRange across python and C++, and it is the only location where _recursive_register_class gets called in python. It itself is mixure of needing to look at python objects and create C++ objects, so the decision of where to put it has previously been arbitrary, but this argues that it is now in the wrong place.

Reply:
I have a hard time debugging python objects in C++, to the point where i'm much less productive. I think if we're going to go in any direction we should move C++ code to python.